service/history/replication/task_store.go

```diff
@@ -0,0 +1,223 @@
+		logger:      logger.WithTags(tag.ComponentReplicationCacheManager),
+		rateLimiter: quotas.NewDynamicRateLimiter(config.ReplicationTaskGenerationQPS.AsFloat64()),
+	}
+}
+
+// Get will return a hydrated replication message for a given cluster based on raw task info.
+// It will either return it immediately from cache or hydrate it, store in cache and then return.
+//
+// Returned task may be nil. This may be due domain not existing in a given cluster or replication message is not longer relevant.
+// Either case is valid and such replication message should be ignored and not returned in the response.
+func (m *TaskStore) Get(ctx context.Context, cluster string, info persistence.ReplicationTaskInfo) (*types.ReplicationTask, error) {
+	cache, ok := m.clusters[cluster]
+	if !ok {
+		return nil, errUnknownCluster
+	}
+
+	domain, err := m.domains.GetDomainByID(info.DomainID)
+	if err != nil {
+		return nil, fmt.Errorf("resolving domain: %w", err)
+	}
+
+	// Domain does not exist in this cluster, do not replicate the task
+	if !domain.HasReplicationCluster(cluster) {
+		return nil, nil
+	}
+
+	scope := m.scope.Tagged(metrics.SourceClusterTag(cluster))
+
+	scope.IncCounter(metrics.CacheRequests)
+	sw := scope.StartTimer(metrics.CacheLatency)
+	defer sw.Stop()
+
+	task := cache.Get(info.TaskID)
+	if task != nil {
+		scope.IncCounter(metrics.CacheHitCounter)
+		return task, nil
+	}
+
+	m.scope.IncCounter(metrics.CacheMissCounter)
+
+	// Rate limit to not kill the database
+	m.rateLimiter.Wait(ctx)
+
+	err = m.throttleRetry.Do(ctx, func() error {
+		var err error
+		task, err = m.hydrator.Hydrate(ctx, info)
+		return err
+	})
+
+	if err != nil {
+		m.scope.IncCounter(metrics.CacheFailures)
+		return nil, err
+	}
+
+	m.Put(task)
+
+	return task, nil
+}
+
+// Put will try to store hydrated replication to all cluster caches.
+// Tasks may not be relevant, as domain is not enabled in some clusters. Ignore task for that cluster.
+// Some clusters may be already have full cache. Ignore task, it will be fetched and hydrated again later.
+// Some clusters may already acknowledged such task. Ignore task, it is no longer relevant for such cluster.
+func (m *TaskStore) Put(task *types.ReplicationTask) {
+	// Do not store nil tasks
+	if task == nil {
+		return
+	}
+
+	domain, err := m.getDomain(task)
+	if err != nil {
+		m.logger.Error("failed to resolve domain", tag.Error(err))
+		return
+	}
+
+	for cluster, cache := range m.clusters {
+		if domain != nil && !domain.HasReplicationCluster(cluster) {
+			continue
+		}
+
+		scope := m.scope.Tagged(metrics.SourceClusterTag(cluster))
+
+		err := cache.Put(task)
+		switch err {
+		case errCacheFull:
+			scope.IncCounter(metrics.CacheFullCounter)
+		case errAlreadyAcked:
+			// No action, this is expected.
+			// Some cluster(s) may be already past this, due to different fetch rates.
+		}
+
+		scope.UpdateGauge(metrics.CacheSize, float64(cache.Size()))
```

Comment:
there's always caveat with gauges. Let's say we have 10 history instances and all of those will report gauge metric with different value. Usually, last one wins, but it will be not meaningful at all.

Add additional, "shard" tag to expose cache size per shard

Reply:
Could m3 tolerate that kind of cardinality? I'm not sure it could. But your point about this being therefore not helpful if it's last-write-wins is a good one

Reply:
It can, but barely. We already have a few places that add shard tag. In particular DLQ size, and replication backlog, fetched tasks, etc. in replication/task_ack_manager.go.